File: ScriptureAloneCore/Sources/ScriptureAloneCore/Interlinear/InterlinearStore.swift
import Foundation
import SQLite3

public enum InterlinearStoreError: Error, LocalizedError {
    case open(String)
    case query(String)
    /// The word records of this verse do not fit the text they were handed — the caller passed a
    /// different translation's verse. The interlinear aligns to the Berean Standard Bible only.
    case textMismatch(VerseRef)
    /// A `BibleStore` that is not the BSB was handed to `words(for:from:)`.
    case wrongTranslation(String)

    public var errorDescription: String? {
        switch self {
        case .open(let message): "Couldn’t open the original-language data: \(message)"
        case .query(let message): "Couldn’t read the original-language data: \(message)"
        case .textMismatch(let verse):
            "The original-language data for \(verse.display) doesn’t match this text. "
            + "Interlinear data is aligned to the Berean Standard Bible."
        case .wrongTranslation(let abbreviation):
            "Interlinear data is aligned to the Berean Standard Bible, not \(abbreviation)."
        }
    }
}

/// Read-only access to the bundled interlinear database (built by `Tools/build_interlinear.py`):
/// word-level Hebrew, Aramaic and Greek for the whole Bible, aligned to the bundled Berean Standard
/// Bible, plus STEPBible's Strong's lexicon.
///
/// Opened `immutable=1`: the file genuinely never changes once bundled, so SQLite takes no locks and
/// caches pages indefinitely. The `NSLock` serializes the one connection, exactly as `StudyStore`
/// does; nothing here touches the main actor.
///
/// **The alignment is to the BSB and only the BSB.** A word's English is not stored — it is a
/// UTF-16 slice of the BSB's own verse text, which is what makes the highlight exact by
/// construction instead of re-deriving the tables' spacing rules at runtime. Handing this store the
/// ASV's or KJV's text for the same verse produces nonsense, so it refuses: see
/// `alignsTo(_:)`, which is the check the UI uses to hide the affordance.
///
/// This compiles on watchOS as it stands (no 64-bit arithmetic, nothing beyond Foundation and
/// SQLite3) and so needs no `#if !os(watchOS)` guard. The watch does not *bundle* the 11 MB
/// database — `ScriptureAloneWatch` lists its resources file by file in `project.yml` and this is
/// not one of them — so on the watch there is simply nothing for it to open.
public final class InterlinearStore: @unchecked Sendable {
    /// The one translation this data aligns to, as `TranslationInfo.id` spells it.
    public static let translationID = "BSB"

    /// Whether a translation can show interlinear data at all. False for the ASV and the KJV, for
    /// every import, and for every online translation — the alignment exists for the BSB alone.
    public static func alignsTo(_ translation: TranslationInfo) -> Bool {
        translation.id == translationID
    }

    public let url: URL
    public let attribution: InterlinearAttribution
    public let statistics: InterlinearStatistics

    private let db: OpaquePointer
    private let lock = NSLock()
    /// 3,821 interned parsing codes, loaded once. Immutable after `init`, so it needs no lock.
    private let parsings: [Int: InterlinearParsing]
    /// Inflated chapter blobs and lexicon buckets, both guarded by `lock`.
    private var chapterCache = InflatedCache<[Substring]>(limit: 4)
    private var lexiconCache = InflatedCache<[Substring]>(limit: 8)

    public init(url: URL) throws {
        self.url = url
        var handle: OpaquePointer?
        let uri = "file:\(url.path(percentEncoded: true))?immutable=1"
        guard sqlite3_open_v2(uri, &handle, SQLITE_OPEN_READONLY | SQLITE_OPEN_URI | SQLITE_OPEN_NOMUTEX, nil) == SQLITE_OK,
              let handle else {
            let message = handle.map { String(cString: sqlite3_errmsg($0)) } ?? "unknown error"
            sqlite3_close(handle)
            throw InterlinearStoreError.open(message)
        }
        db = handle

        var meta: [String: String] = [:]
        try Self.rows(db, "SELECT key, value FROM meta") { stmt in
            meta[Self.string(stmt, 0)] = Self.string(stmt, 1)
        }
        attribution = InterlinearAttribution(
            words: meta["bsb_attribution"] ?? "",
            wordsLicense: meta["bsb_license"] ?? "",
            wordsLicenseURL: URL(string: meta["bsb_license_url"] ?? ""),
            wordsSourceURL: URL(string: meta["bsb_url"] ?? ""),
            lexicon: meta["stepbible_attribution"] ?? "",
            lexiconLicense: meta["stepbible_license"] ?? "",
            lexiconLicenseURL: URL(string: meta["stepbible_license_url"] ?? ""),
            lexiconSourceURL: URL(string: meta["stepbible_url"] ?? ""),
            lexiconChanges: meta["stepbible_changes"] ?? "")
        statistics = InterlinearStatistics(
            version: meta["version"] ?? "",
            checked: meta["checked"] ?? "",
            lexiconCommit: meta["stepbible_commit"] ?? "",
            words: Int(meta["words"] ?? "") ?? 0,
            taggedWords: Int(meta["tagged_words"] ?? "") ?? 0,
            strongsNumbers: Int(meta["strongs"] ?? "") ?? 0,
            versesAligned: Int(meta["verses_aligned"] ?? "") ?? 0,
            verses: Int(meta["verses_total"] ?? "") ?? 0)

        var codes: [Int: InterlinearParsing] = [:]
        try Self.rows(db, "SELECT id, code, description FROM parsings") { stmt in
            codes[Int(sqlite3_column_int64(stmt, 0))] = InterlinearParsing(code: Self.string(stmt, 1),
                                                                          description: Self.string(stmt, 2))
        }
        parsings = codes
    }

    deinit { sqlite3_close(db) }

    // MARK: - Coverage

    /// How many original-language words each verse of the chapter has. Verses with none are absent,
    /// so `counts[verse.key] != nil` is the answer to "does this verse have interlinear data".
    ///
    /// This is the call to make while rendering a chapter: one indexed range scan over `verses`
    /// (~30 rows) and no blob is inflated. `crossReferenceCounts(in:)` has the same shape.
    public func wordCounts(in chapter: ChapterRef) throws -> [Int: Int] {
        try locked {
            var counts: [Int: Int] = [:]
            try Self.rows(db, """
                SELECT verse_key, count FROM verses
                WHERE verse_key BETWEEN ?1 AND ?2 AND count > 0
                """, bind: [chapter.keyRange.lowerBound, chapter.keyRange.upperBound]) { stmt in
                counts[Int(sqlite3_column_int64(stmt, 0))] = Int(sqlite3_column_int64(stmt, 1))
            }
            return counts
        }
    }

    /// Whether one verse has any original-language words. One primary-key lookup, no inflation.
    ///
    /// Nehemiah 7:68 is the only verse of the BSB with none — the BSB supplies it from "some Hebrew
    /// manuscripts" and most of the Masoretic Text does not have it.
    public func hasWords(for verse: VerseRef) throws -> Bool {
        try locked {
            var count = 0
            try Self.rows(db, "SELECT count FROM verses WHERE verse_key = ?1", bind: [verse.key]) { stmt in
                count = Int(sqlite3_column_int64(stmt, 0))
            }
            return count > 0
        }
    }

    // MARK: - Words

    /// Every original-language word of one verse, in English (BSB) reading order, with the English
    /// each one renders already resolved.
    ///
    /// A verse with no data — Nehemiah 7:68, or any key this database does not carry — returns an
    /// empty array rather than throwing. Superscription words (Psalm titles, Zechariah 12:1's
    /// oracle heading) come first and are flagged; the BSB keeps them out of `verseText`.
    ///
    /// - Parameters:
    ///   - verse: the verse to read.
    ///   - verseText: **the BSB's** text for that verse, exactly as `BibleStore` stores it. The
    ///     caller supplies it rather than this store fetching it, for two reasons: the UI already
    ///     has that string on screen — it is the very text the returned `range`s highlight, so
    ///     re-reading it here would risk highlighting a *different* string than the one displayed —
    ///     and an interlinear store that owned a `BibleStore` would own a translation's lifetime,
    ///     which is the reader's job, not the study data's. Pass the wrong translation's text and
    ///     this throws `textMismatch` rather than silently mis-highlighting; the
    ///     `words(for:from:)` overload does the fetching when that is what you want.
    public func words(for verse: VerseRef, in verseText: String) throws -> [InterlinearWord] {
        let records = try locked { try self.records(for: verse) }
        guard !records.isEmpty else { return [] }
        let units = Array(verseText.utf16)
        var result: [InterlinearWord] = []
        result.reserveCapacity(records.count)
        for (position, record) in records.enumerated() {
            let fields = record.split(separator: "\t", omittingEmptySubsequences: false)
            guard fields.count >= 10 else {
                throw InterlinearStoreError.query("malformed word record in \(verse.display)")
            }
            let start = Int(fields[7]) ?? -1
            let length = Int(fields[8]) ?? -1
            var english = String(fields[0])
            var range: NSRange?
            if start >= 0, length >= 0 {
                guard start + length <= units.count else { throw InterlinearStoreError.textMismatch(verse) }
                english = String(decoding: units[start..<(start + length)], as: UTF16.self)
                range = NSRange(location: start, length: length)
            }
            let strongs = String(fields[4])
            let parsingID = Int(fields[3]) ?? 0
            result.append(InterlinearWord(
                position: position,
                english: english,
                original: String(fields[1]),
                transliteration: String(fields[2]),
                parsing: parsings[parsingID],
                strongs: strongs.isEmpty ? nil : strongs,
                language: InterlinearLanguage(rawValue: String(fields[5])) ?? .hebrew,
                originalOrder: Int(fields[6]) ?? position + 1,
                range: range,
                isSuperscription: (Int(fields[9]) ?? 0) & 1 == 1))
        }
        return result
    }

    /// `words(for:in:)` with the verse text read out of a `BibleStore`, which must be the BSB.
    public func words(for verse: VerseRef, from bible: BibleStore) throws -> [InterlinearWord] {
        guard Self.alignsTo(bible.info) else {
            throw InterlinearStoreError.wrongTranslation(bible.info.abbreviation.isEmpty ? bible.info.id
                                                                                         : bible.info.abbreviation)
        }
        let text = try bible.verses(in: VerseRange(verse)).first?.text ?? ""
        return try words(for: verse, in: text)
    }

    /// The verse's raw records, sliced out of its chapter blob. Caller holds `lock`.
    private func records(for verse: VerseRef) throws -> ArraySlice<Substring> {
        var chapterKey = 0, first = 0, count = 0
        try Self.rows(db, "SELECT chapter_key, first, count FROM verses WHERE verse_key = ?1",
                      bind: [verse.key]) { stmt in
            chapterKey = Int(sqlite3_column_int64(stmt, 0))
            first = Int(sqlite3_column_int64(stmt, 1))
            count = Int(sqlite3_column_int64(stmt, 2))
        }
        guard count > 0 else { return [] }
        let lines = try chapter(chapterKey)
        guard first >= 0, first + count <= lines.count else {
            throw InterlinearStoreError.query("chapter \(chapterKey) is shorter than \(verse.display) claims")
        }
        return lines[first..<(first + count)]
    }

    /// One chapter's word records, inflated at most once per cache eviction. Caller holds `lock`.
    private func chapter(_ chapterKey: Int) throws -> [Substring] {
        if let cached = chapterCache[chapterKey] { return cached }
        var body = Data()
        try Self.rows(db, "SELECT words FROM chapters WHERE chapter_key = ?1", bind: [chapterKey]) { stmt in
            body = Self.blob(stmt, 0)
        }
        guard let text = StudyStore.inflate(body) else {
            throw InterlinearStoreError.query("chapter \(chapterKey) does not inflate")
        }
        let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
        chapterCache[chapterKey] = lines
        return lines
    }

    // MARK: - Lexicon

    /// The Strong's number's lexicon entry: gloss, fuller definition and sub-senses in order.
    ///
    /// Accepts the zero-padded form the word records use ("H0430") and the way people write it
    /// ("H430", "h430"); an extended key's suffix is dropped, because the build collapsed extended
    /// keys into their base number. Returns nil for a number no entry covers.
    public func entry(for strongs: String) throws -> LexiconEntry? {
        guard let number = Self.normalize(strongs) else { return nil }
        return try locked {
            var bucketID = 0, part = -1
            try Self.rows(db, "SELECT entry_id, part FROM lexicon WHERE strongs = ?1", bind: [number]) { stmt in
                bucketID = Int(sqlite3_column_int64(stmt, 0))
                part = Int(sqlite3_column_int64(stmt, 1))
            }
            guard part >= 0 else { return nil }
            let bucket = try self.lexiconBucket(bucketID)
            guard part < bucket.count else {
                throw InterlinearStoreError.query("lexicon bucket \(bucketID) has no part \(part)")
            }
            // Senses are separated by U+001E, their five fields by tabs.
            let senses = bucket[part].split(separator: "\u{1e}", omittingEmptySubsequences: false).compactMap { sense -> LexiconSense? in
                let fields = sense.split(separator: "\t", omittingEmptySubsequences: false)
                guard fields.count >= 5 else { return nil }
                return LexiconSense(lemma: String(fields[0]), transliteration: String(fields[1]),
                                    morphology: String(fields[2]), gloss: String(fields[3]),
                                    definition: String(fields[4]))
            }
            return senses.isEmpty ? nil : LexiconEntry(strongs: number, senses: senses)
        }
    }

    /// A bucket of 64 entries, separated by U+0000. Caller holds `lock`.
    private func lexiconBucket(_ id: Int) throws -> [Substring] {
        if let cached = lexiconCache[id] { return cached }
        var body = Data()
        try Self.rows(db, "SELECT body FROM lexicon_text WHERE id = ?1", bind: [id]) { stmt in
            body = Self.blob(stmt, 0)
        }
        guard let text = StudyStore.inflate(body) else {
            throw InterlinearStoreError.query("lexicon bucket \(id) does not inflate")
        }
        let parts = text.split(separator: "\0", omittingEmptySubsequences: false)
        lexiconCache[id] = parts
        return parts
    }

    /// "h430" and "H0430G" both mean the entry stored as "H0430". Nil for anything else.
    static func normalize(_ strongs: String) -> String? {
        var letter: Character?
        var digits = ""
        for character in strongs.trimmingCharacters(in: .whitespaces) {
            if letter == nil {
                guard character == "H" || character == "h" || character == "G" || character == "g" else { return nil }
                letter = Character(character.uppercased())
            } else if character.isNumber {
                guard digits.count < 5 else { return nil }
                digits.append(character)
            } else if character.isLetter {
                break   // an extended key's suffix: H0430G is stored under H0430
            } else {
                return nil
            }
        }
        guard let letter, let value = Int(digits) else { return nil }
        return "\(letter)\(String(format: "%04d", value))"
    }

    // MARK: - Cache

    /// A handful of inflated blobs, evicted oldest-first. A reader moves through one chapter at a
    /// time and a lexicon bucket holds 64 consecutive Strong's numbers, so a very small cache turns
    /// a verse's worth of taps into one inflation instead of one per word.
    private struct InflatedCache<Value> {
        private let limit: Int
        private var order: [Int] = []
        private var values: [Int: Value] = [:]

        init(limit: Int) { self.limit = limit }

        subscript(key: Int) -> Value? {
            get { values[key] }
            set {
                guard let newValue else {
                    values[key] = nil
                    order.removeAll { $0 == key }
                    return
                }
                if values[key] == nil {
                    order.append(key)
                    if order.count > limit { values[order.removeFirst()] = nil }
                }
                values[key] = newValue
            }
        }
    }

    // MARK: - SQLite helpers

    private func locked<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    private static func rows(_ db: OpaquePointer, _ sql: String, bind: [Int] = [], _ row: (OpaquePointer) throws -> Void) throws {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt else {
            throw InterlinearStoreError.query(String(cString: sqlite3_errmsg(db)))
        }
        defer { sqlite3_finalize(stmt) }
        for (index, value) in bind.enumerated() {
            sqlite3_bind_int64(stmt, Int32(index + 1), sqlite3_int64(value))
        }
        try step(db, stmt, row)
    }

    private static func rows(_ db: OpaquePointer, _ sql: String, bind: [String], _ row: (OpaquePointer) throws -> Void) throws {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt else {
            throw InterlinearStoreError.query(String(cString: sqlite3_errmsg(db)))
        }
        defer { sqlite3_finalize(stmt) }
        let transient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
        for (index, value) in bind.enumerated() {
            sqlite3_bind_text(stmt, Int32(index + 1), value, -1, transient)
        }
        try step(db, stmt, row)
    }

    private static func step(_ db: OpaquePointer, _ stmt: OpaquePointer, _ row: (OpaquePointer) throws -> Void) throws {
        while true {
            let status = sqlite3_step(stmt)
            if status == SQLITE_ROW {
                try row(stmt)
            } else if status == SQLITE_DONE {
                return
            } else {
                throw InterlinearStoreError.query(String(cString: sqlite3_errmsg(db)))
            }
        }
    }

    private static func string(_ stmt: OpaquePointer, _ column: Int32) -> String {
        guard let cString = sqlite3_column_text(stmt, column) else { return "" }
        return String(cString: cString)
    }

    private static func blob(_ stmt: OpaquePointer, _ column: Int32) -> Data {
        let count = Int(sqlite3_column_bytes(stmt, column))
        guard count > 0, let pointer = sqlite3_column_blob(stmt, column) else { return Data() }
        return Data(bytes: pointer, count: count)
    }
}

File: android/app/src/main/java/com/blainemiller/scripturealone/data/study/InterlinearStore.kt
package com.blainemiller.scripturealone.data.study

import com.blainemiller.scripturealone.data.sql.SqlSource

/** The language one interlinear word is written in. */
enum class InterlinearLanguage(val code: String, val displayName: String) {
    HEBREW("H", "Hebrew"), ARAMAIC("A", "Aramaic"), GREEK("G", "Greek");

    /** Hebrew and Aramaic are written right to left; Greek is not. */
    val isRightToLeft: Boolean get() = this != GREEK

    companion object {
        fun of(code: String): InterlinearLanguage = entries.firstOrNull { it.code == code } ?: HEBREW
    }
}

/** A morphological parsing, interned in the database's `parsings` table. */
data class InterlinearParsing(val code: String, val description: String)

/**
 * One original-language word of a verse, with the English the BSB renders it as.
 *
 * [english] can legitimately be empty — Hebrew's object marker, a Greek article the BSB folds into
 * the following noun. That is a word to show with a blank English cell, not one to drop.
 *
 * [range] is where [english] sits in the BSB verse text, in UTF-16 units: exactly Kotlin's string
 * indices, so it slices and highlights directly. Null for a superscription word (the BSB keeps Psalm
 * titles out of the verse), a word rendered with nothing, or a word of one of the few verses whose
 * records don't reproduce the BSB text exactly.
 */
data class InterlinearWord(
    val position: Int,
    val english: String,
    val original: String,
    val transliteration: String,
    val parsing: InterlinearParsing?,
    val strongs: String?,
    val language: InterlinearLanguage,
    /** 1-based position in the original-language sentence; over a verse, a permutation of 1…n. */
    val originalOrder: Int,
    val range: IntRange?,
    val isSuperscription: Boolean,
) {
    val parsingDescription: String get() = parsing?.description ?: ""
}

/** One sense of a Strong's number, as STEPBible's TBESH / TBESG write it. */
data class LexiconSense(
    val lemma: String,
    val transliteration: String,
    val morphology: String,
    val gloss: String,
    val definition: String,
) {
    val lines: List<String> get() = definition.split('\n').filter { it.isNotEmpty() }
}

/** A Strong's number's entry: its senses in STEPBible's order. */
data class LexiconEntry(val strongs: String, val senses: List<LexiconSense>) {
    val gloss: String get() = senses.firstOrNull()?.gloss ?: ""
    val lemma: String get() = senses.firstOrNull()?.lemma ?: ""
    val transliteration: String get() = senses.firstOrNull()?.transliteration ?: ""
    /** The fullest definition the entry carries — the first sense that has one. */
    val definition: String get() = senses.firstOrNull { it.definition.isNotEmpty() }?.definition ?: ""
}

/**
 * The attribution both sources ask for. STEPBible's CC BY 4.0 *requires* credit and a statement of
 * changes, so [requiredLines] must appear somewhere the reader can reach.
 */
data class InterlinearAttribution(
    val words: String,
    val wordsLicense: String,
    val wordsLicenseUrl: String,
    val wordsSourceUrl: String,
    val lexicon: String,
    val lexiconLicense: String,
    val lexiconLicenseUrl: String,
    val lexiconSourceUrl: String,
    val lexiconChanges: String,
) {
    val requiredLines: List<String> get() = listOf(words, lexicon, lexiconChanges)
}

/** What the database says about itself — the counts the build recorded. */
data class InterlinearStatistics(
    val version: String,
    val checked: String,
    val lexiconCommit: String,
    val words: Int,
    val taggedWords: Int,
    val strongsNumbers: Int,
    val versesAligned: Int,
    val verses: Int,
)

/** The verse's records don't fit the text passed in — it isn't the BSB's text for that verse. */
class InterlinearTextMismatch(verseKey: Int) :
    IllegalArgumentException("Original-language data for $verseKey doesn't match this text; it is aligned to the BSB.")

/**
 * Word-level Hebrew, Aramaic and Greek for the whole Bible, aligned to the bundled BSB, plus
 * STEPBible's Strong's lexicon. A port of `ScriptureAloneCore/Interlinear/InterlinearStore.swift`.
 *
 * **The alignment is to the BSB's text.** A word's English is a slice of the BSB verse, so
 * [words] must be given the BSB's text for the verse — the reader offers the original languages in
 * every translation, but always glosses with the BSB. Given another translation's text it throws
 * [InterlinearTextMismatch] rather than mis-highlighting.
 */
class InterlinearStore(private val db: SqlSource) {

    val attribution: InterlinearAttribution
    val statistics: InterlinearStatistics
    private val parsings: Map<Long, InterlinearParsing>
    private val chapterCache = LruCache<Long, List<String>>(4)
    private val lexiconCache = LruCache<Long, List<String>>(8)

    init {
        val meta = db.query("SELECT key, value FROM meta") { it.text(0) to it.text(1) }.toMap()
        fun m(key: String) = meta[key] ?: ""
        fun n(key: String) = meta[key]?.toIntOrNull() ?: 0
        attribution = InterlinearAttribution(
            m("bsb_attribution"), m("bsb_license"), m("bsb_license_url"), m("bsb_url"),
            m("stepbible_attribution"), m("stepbible_license"), m("stepbible_license_url"),
            m("stepbible_url"), m("stepbible_changes"),
        )
        statistics = InterlinearStatistics(
            m("version"), m("checked"), m("stepbible_commit"), n("words"), n("tagged_words"),
            n("strongs"), n("verses_aligned"), n("verses_total"),
        )
        parsings = db.query("SELECT id, code, description FROM parsings") {
            it.long(0) to InterlinearParsing(it.text(1), it.text(2))
        }.toMap()
    }

    /** Words per verse of a chapter; verses with none are absent. One range scan, nothing inflated. */
    fun wordCounts(book: Int, chapter: Int): Map<Int, Int> {
        val base = book * 1_000_000 + chapter * 1_000
        return db.query(
            "SELECT verse_key, count FROM verses WHERE verse_key BETWEEN ? AND ? AND count > 0",
            base, base + 999,
        ) { it.long(0).toInt() to it.long(1).toInt() }.toMap()
    }

    /** Whether one verse has any words. Nehemiah 7:68 is the only BSB verse with none. */
    fun hasWords(verseKey: Int): Boolean =
        (db.query("SELECT count FROM verses WHERE verse_key = ?", verseKey) { it.long(0) }.firstOrNull() ?: 0L) > 0

    /**
     * Every original-language word of one verse, in English (BSB) reading order. Superscription
     * words come first and are flagged. A verse with no data returns an empty list.
     *
     * @param verseText the **BSB's** text for the verse, exactly as stored — the very string the
     *   returned ranges index into.
     */
    fun words(verseKey: Int, verseText: String): List<InterlinearWord> {
        val records = records(verseKey)
        return records.mapIndexed { position, record ->
            val fields = record.split('\t')
            require(fields.size >= 10) { "malformed word record in $verseKey" }
            val start = fields[7].toIntOrNull() ?: -1
            val length = fields[8].toIntOrNull() ?: -1
            var english = fields[0]
            var range: IntRange? = null
            if (start >= 0 && length >= 0) {
                if (start + length > verseText.length) throw InterlinearTextMismatch(verseKey)
                english = verseText.substring(start, start + length)
                range = start until start + length
            }
            val strongs = fields[4]
            InterlinearWord(
                position = position,
                english = english,
                original = fields[1],
                transliteration = fields[2],
                parsing = parsings[fields[3].toLongOrNull() ?: 0L],
                strongs = strongs.ifEmpty { null },
                language = InterlinearLanguage.of(fields[5]),
                originalOrder = fields[6].toIntOrNull() ?: (position + 1),
                range = range,
                isSuperscription = ((fields[9].toIntOrNull() ?: 0) and 1) == 1,
            )
        }
    }

    private fun records(verseKey: Int): List<String> {
        data class Row(val chapterKey: Long, val first: Int, val count: Int)
        val row = db.query("SELECT chapter_key, first, count FROM verses WHERE verse_key = ?", verseKey) {
            Row(it.long(0), it.long(1).toInt(), it.long(2).toInt())
        }.firstOrNull() ?: return emptyList()
        if (row.count <= 0) return emptyList()
        val lines = chapter(row.chapterKey)
        check(row.first >= 0 && row.first + row.count <= lines.size) {
            "chapter ${row.chapterKey} is shorter than verse $verseKey claims"
        }
        return lines.subList(row.first, row.first + row.count)
    }

    private fun chapter(chapterKey: Long): List<String> = chapterCache.getOrPut(chapterKey) {
        val body = db.query("SELECT words FROM chapters WHERE chapter_key = ?", chapterKey) { it.blob(0) }
            .firstOrNull() ?: ByteArray(0)
        val text = StudyStore.inflate(body) ?: error("chapter $chapterKey does not inflate")
        text.split('\n')
    }

    /**
     * A Strong's number's lexicon entry. Accepts "H0430" as the records spell it and "H430"/"h430"
     * as people write it; an extended key's suffix is dropped, as the build collapsed them.
     */
    fun entry(strongs: String): LexiconEntry? {
        val number = normalize(strongs) ?: return null
        data class Row(val bucket: Long, val part: Int)
        val row = db.query("SELECT entry_id, part FROM lexicon WHERE strongs = ?", number) {
            Row(it.long(0), it.long(1).toInt())
        }.firstOrNull() ?: return null
        val bucket = lexiconBucket(row.bucket)
        check(row.part < bucket.size) { "lexicon bucket ${row.bucket} has no part ${row.part}" }
        // Senses are separated by U+001E, their five fields by tabs.
        val senses = bucket[row.part].split(SENSE_SEPARATOR).mapNotNull { sense ->
            val f = sense.split('\t')
            if (f.size < 5) null else LexiconSense(f[0], f[1], f[2], f[3], f[4])
        }
        return if (senses.isEmpty()) null else LexiconEntry(number, senses)
    }

    /** A bucket of 64 entries, separated by U+0000. */
    private fun lexiconBucket(id: Long): List<String> = lexiconCache.getOrPut(id) {
        val body = db.query("SELECT body FROM lexicon_text WHERE id = ?", id) { it.blob(0) }
            .firstOrNull() ?: ByteArray(0)
        val text = StudyStore.inflate(body) ?: error("lexicon bucket $id does not inflate")
        text.split(ENTRY_SEPARATOR)
    }

    /**
     * A handful of inflated blobs, evicted oldest-first. A reader moves through one chapter at a time
     * and a lexicon bucket holds 64 consecutive numbers, so a small cache turns a verse's worth of
     * taps into one inflation.
     */
    private class LruCache<K, V>(private val limit: Int) {
        private val map = LinkedHashMap<K, V>()

        @Synchronized
        fun getOrPut(key: K, compute: () -> V): V {
            map[key]?.let { return it }
            val value = compute()
            map[key] = value
            if (map.size > limit) map.remove(map.keys.first())
            return value
        }
    }

    companion object {
        private const val SENSE_SEPARATOR = ''
        private const val ENTRY_SEPARATOR = ' '

        /** "h430" and "H0430G" both mean the entry stored as "H0430". Null for anything else. */
        fun normalize(strongs: String): String? {
            var letter: Char? = null
            val digits = StringBuilder()
            for (c in strongs.trim()) {
                if (letter == null) {
                    if (c !in "HhGg") return null
                    letter = c.uppercaseChar()
                } else if (c.isDigit()) {
                    if (digits.length >= 5) return null
                    digits.append(c)
                } else if (c.isLetter()) {
                    break // an extended key's suffix: H0430G is stored under H0430
                } else {
                    return null
                }
            }
            val value = digits.toString().toIntOrNull() ?: return null
            return "${letter ?: return null}${value.toString().padStart(4, '0')}"
        }
    }
}
